pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -0,0 +1,720 @@
+	// NewController.
+	clustersLister capilisters.ClusterLister
+	// clustersSynced returns true if the cluster shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	clustersSynced cache.InformerSynced
+
+	// machineSetsLister is able to list/get machine sets and is populated by the shared informer passed to
+	// NewController.
+	machineSetsLister capilisters.MachineSetLister
+	// machineSetsSynced returns true if the machine set shared informer has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	machineSetsSynced cache.InformerSynced
+
+	// clusterVersionsLister is able to list/get clusterversions and is populated by the shared
+	// informer passed to NewClusterController.
+	clusterVersionsLister clustoplisters.ClusterVersionLister
+
+	// Clusters that need to be synced
+	queue workqueue.RateLimitingInterface
+
+	// logger for controller
+	logger log.FieldLogger
+}
+
+func (c *Controller) addClusterDeployment(obj interface{}) {
+	clusterDeployment := obj.(*clustop.ClusterDeployment)
+	clustoplog.WithClusterDeployment(c.logger, clusterDeployment).Debugf("adding cluster deployment")
+	c.enqueueClusterDeployment(clusterDeployment)
+}
+
+func (c *Controller) updateClusterDeployment(old, cur interface{}) {
+	oldClusterDeployment := old.(*clustop.ClusterDeployment)
+	curClusterDeployment := cur.(*clustop.ClusterDeployment)
+	clustoplog.WithClusterDeployment(c.logger, oldClusterDeployment).Debugf("updating cluster deployment")
+	c.enqueueClusterDeployment(curClusterDeployment)
+}
+
+func (c *Controller) deleteClusterDeployment(obj interface{}) {
+	clusterDeployment, ok := obj.(*clustop.ClusterDeployment)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Couldn't get object from tombstone %#v", obj))
+			return
+		}
+		clusterDeployment, ok = tombstone.Obj.(*clustop.ClusterDeployment)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("Tombstone contained object that is not a Cluster Deployment %#v", obj))
+			return
+		}
+	}
+	clustoplog.WithClusterDeployment(c.logger, clusterDeployment).Debugf("deleting cluster deployment")
+	c.enqueueClusterDeployment(clusterDeployment)
+}
+
+// When a cluster-api cluster is created, enqueue the cluster deployment that owns it and update its expectations.
+func (c *Controller) addCluster(obj interface{}) {
+	cluster := obj.(*capi.Cluster)
+
+	if cluster.DeletionTimestamp != nil {
+		// on a restart of the controller manager, it's possible a cluster shows up in a state that
+		// is already pending deletion. Prevent the cluster from being a creation observation.
+		c.deleteCluster(cluster)
+		return
+	}
+
+	clusterDeployment, err := controller.ClusterDeploymentForCluster(cluster, c.clusterDeploymentsLister)
+	if err != nil {
+		clustoplog.WithCluster(c.logger, cluster).Errorf("error retrieving cluster deployment for cluster: %v", err)
```

Comment:
This should only be an error if the Cluster has a clust-op ProviderConfig. We should ignore Clusters intended for other providers.

Reply:
We pass the cluster kind when retrieving the owner, so if an error happens, it's because the resource did have a valid owner reference to a clusterdeployment, and something happened retrieving it.

Reply:
Good point.